### PR TITLE
Avoid returning the same entity multiple times from separate searches

### LINKED ISSUE #15604: Display a duplicate user when we use ldap federation
### Area

ldap

### Describe the bug

1 - Configure ldap provider to use federation user (openldap)
    Synchronization settings
       Import users  Off
       Sync Registrations Off
   Batch size 
       Periodic full sync Off
       Periodic changed users sync  Off
2 - When i create the first user in my ldap. the search user in keycloak UI return duplicate user (the same use)  ==> KO
3 - When i create the seconde user in my ldap the search user in keycloak UI return only one (not duplicate) ==> OK

### Version

20.0.1

### Expected behavior

We need get only one result for the first user creation in ldap

### Actual behavior

returned two result for the first user creation in ldap

### How to Reproduce?

1 - Configure ldap provider to use federation user (openldap)
    Synchronization settings
       Import users  Off
       Sync Registrations Off
   Batch size 
       Periodic full sync Off
       Periodic changed users sync  Off
2 - When i create the first user in my ldap. the search user in keycloak UI return duplicate user (the same use)  ==> KO
3 - When i create the seconde user in my ldap the search user in keycloak UI return only one (not duplicate) ==> OK

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by ahus1**
Closes #15604

### CODE REVIEW

File: federation/ldap/src/main/java/org/keycloak/storage/ldap/LDAPStorageProvider.java [hunk 1]
```diff
@@ -459,7 +459,12 @@ protected List<LDAPObject> searchLDAP(RealmModel realm, Map<String, String> attr
                 ldapQuery.addWhereCondition(emailCondition);
 
                 List<LDAPObject> ldapObjects = ldapQuery.getResultList();
-                results.addAll(ldapObjects);
+                ldapObjects.forEach(ldapObject -> {
+                    // ensure that no entity is listed twice and still preserve the order of returned entities
+                    if (!results.contains(ldapObject)) {
+                        results.addAll(ldapObjects);
```
Comment (hmlnarik) [inline federation/ldap/src/main/java/org/keycloak/storage/ldap/LDAPStorageProvider.java:465]:
```suggestion
                        results.add(ldapObject);
```
(same below)

Comment (hmlnarik) [reply federation/ldap/src/main/java/org/keycloak/storage/ldap/LDAPStorageProvider.java:465]: I am worried of the complexity. Could we instead use a set of unique identifiers to check if the `ldapObject` has already been added?

Comment (ahus1) [reply federation/ldap/src/main/java/org/keycloak/storage/ldap/LDAPStorageProvider.java]: @hmlnarik - reworked. Still, if we would have computational complexity here, I assume the result would be too big for the caller anyway / would take too long to retrieve from LDAP.

**Review by hmlnarik — COMMENTED** [review]
Thank you @ahus1. The code looks good to me now.

The only missing piece is test. Is there a reason to omit it?

File: federation/ldap/src/main/java/org/keycloak/storage/ldap/LDAPStorageProvider.java (hunk 1 above)
Comment (hmlnarik) [reply federation/ldap/src/main/java/org/keycloak/storage/ldap/LDAPStorageProvider.java]: Thank you, resolving this

**Comment by ahus1** [comment]
@hmlnarik - I'm just starting to look into the tests. I assume that you noticed this PR due to the codeowners notification and I should have created this PR as a draft first.

**Comment by hmlnarik** [comment]
@ahus1 Yes, precisely. Draft state helps, thank you

**Comment by ahus1** [comment]
@hmlnarik - I've now added tests. Ready for review.

**Comment by ahus1** [comment]
@hmlnarik - some unstable tests reported just when I switched it to ready for review. Reverting for now as I need to check why this has been reported.

**Comment by ahus1** [comment]
@hmlnarik - Reworked the tests, should no longer be unstable - unstability was caused by the same user name used in multiple tests. Ready for review.

**Review by sguilhen — APPROVED** [review]

**Review by hmlnarik — APPROVED** [review]
Thank you

**Pull request merged**